corehq/util/couch.py

```diff
@@ -33,3 +36,130 @@ def categorize_bulk_save_errors(error):
         result_map[error].append(result)
 
     return result_map
+
+
+class IterDB(object):
+    """
+    Bulk save docs in chunks.
+
+        with IterDB(db) as iter_db:
+            for doc in iter_docs(db, ids):
+                iter_db.save(doc)
+
+        iter_db.error_ids  # docs that errored
+        iter_db.saved_ids  # docs that saved correctly
+    """
+    def __init__(self, database, chunksize=100, throttle_secs=None):
+        self.db = database
+        self.chunksize = chunksize
+        self.throttle_secs = throttle_secs
+        self.saved_ids = []
+        self.deleted_ids = []
+        self.error_ids = []
+
+    def __enter__(self):
+        self.to_save = []
+        self.to_delete = []
+        return self
+
+    def commit(self, cmd, docs):
+        try:
+            results = cmd(docs)
+        except BulkSaveError as e:
+            results = e.results
+            error_ids = {d['id'] for d in e.errors}
```

Comment:
I guess `BulkSaveError` is inherently not that helpful, but is there anything more we could do to be smarter about what these errors are caused by, especially since we are going to keep trying updating these error_ids?